From e60c5608f843ec106a98a98b33de0c3be070d557 Mon Sep 17 00:00:00 2001
From: Arif Hussain <arifhussain@codeaurora.org>
Date: Fri, 11 May 2018 06:59:27 -0700
Subject: qcacld-3.0: Avoid int overflow in
 csr_scan_save_preferred_network_found()

Add validation check on frameLength to avoid int overflow in
csr_scan_save_preferred_network_found function.

Change-Id: I0f2a0557fa60e81f0b9d003ae73091f2974046e8
CRs-Fixed: 2232358
---
 core/sme/src/csr/csr_api_scan.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/core/sme/src/csr/csr_api_scan.c b/core/sme/src/csr/csr_api_scan.c
index 0d5aed3..f9972a4 100644
--- a/core/sme/src/csr/csr_api_scan.c
+++ b/core/sme/src/csr/csr_api_scan.c
@@ -7813,6 +7813,13 @@ QDF_STATUS csr_scan_save_preferred_network_found(tpAniSirGlobal pMac,
 		uLen = pPrefNetworkFoundInd->frameLength -
 		       (SIR_MAC_HDR_LEN_3A + SIR_MAC_B_PR_SSID_OFFSET);
 	}
+
+	if (uLen > (UINT_MAX - sizeof(struct tag_csrscan_result))) {
+		sme_err("Incorrect len: %d, may leads to int overflow, uLen %d",
+			pPrefNetworkFoundInd->frameLength, uLen);
+		qdf_mem_free(parsed_frm);
+		return QDF_STATUS_E_FAILURE;
+	}
 	pScanResult = qdf_mem_malloc(sizeof(struct tag_csrscan_result) + uLen);
 	if (NULL == pScanResult) {
 		sme_err("fail to allocate memory for frame");
-- 
cgit v1.1

